lib/ExpensiMark.js

```diff
@@ -346,6 +348,40 @@ export default class ExpensiMark {
         return replacedText;
     }
 
+    /**
+     * Map the block to newline
+     *
+     * @param {String} htmlString
+     *
+     * @returns {String}
+     */
+    mapBlock(htmlString) {
+        const splitText = htmlString.split('[block]');
```

Comment:
@ahmdshrif how about we store indices where a new line needs to be added instead?

Then function's signature could look like `insertNewLine(htmlString, indices)`. 
Thoughts?

https://user-images.githubusercontent.com/29673073/161436076-08677e0a-51c5-4d66-9d79-082adf734ba4.mov

Reply:
@rushatgabhane 
I will expect some issues with this approach like extra lines.  and will need more complex code to fix it 
this issue is something I expect. but if we need to solve it we will add a random number after block. and make sure it's not on the text like 
some pseudo code:
```ts
splitTag = [block]
while (splitTag in text  ){
splitTag = [{block+math.rondome..}]
}
text.replace(tages, splitTag)
mapBlock (text,splitTag)
}

Reply:
How about we use something like `String.matchAll()` or `String.exec()` to get an array of indices.
And then use `String.slice()` to split strings using those indices. Rest of your logic will remain the same.

References 
- String.matchAll - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/matchAll
- String.slice - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/slice

Reply:
@rushatgabhane I don't get exactly what you mean 
here is an example with  result 

```ts
array = [...String.matchAll(regex)]
// reustl : Array [Array ["<div>"], Array ["<div>"], Array ["<div>"]]

```
how we will use slice with this data .

Reply:
> Each match is an array with extra properties index, input.

@ahmdshrif I figure you could use `array[i].index`.
Anyway, we also have [exec()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec). Maybe give that a try.

Reply:
ok, but I know now where the undefined comes from and I fix it. 

```
        joinedText += splitText[splitText.length - 1];
```

when we only have a [block] the return array from the split will be empty so the length will be 0 and length -1  = -1 so it will give undefined. 

now I fix the by adding it as a condition on forEach. 

I don't think we need now to slice the array.

Reply:
@rushatgabhane  what do you think?

Reply:
I'm not a huge fan of it because it feels like a workaround for not using index.
Maybe I'm tunnel visioned, and this is one good way to approach the problem.

Reply:
Ok i think now i get your point . And i will implement it . 

My point was :-
When we use slice it cost o(n)  
And we will slice n time so it will cost o(n^2)

Splite can do this on o(n) time complexity.

Reply:
@ahmdshrif fantastic!

Reply:
@rushatgabhane  finaly I pass all test cases  :D 

for me, we do the code more complex with slow algorism O(N^2).  and the same result  ( maybe I miss understand your idea )
this is the final code do you think we should push it or keep the old one. ? 


```ts
       mapBlock(htmlString) {
        const regex = RegExp('\\[block\\]', 'g');
        // eslint-disable-next-line no-unused-vars
        let array;
        const indexes = [];
        // eslint-disable-next-line no-cond-assign
        while ((array = regex.exec(htmlString)) !== null) {
            indexes.push(regex.lastIndex);
        }
        let joinedText = '';

        // delete empty spaces at the end
        while (indexes.length >= 1) {
            const string = htmlString.substring(indexes[indexes.length - 2], indexes[indexes.length - 1] - 7);
            if (string.trim().length > 0 || string.match(/\n/)) {
                break;
            }
            indexes.pop();
        }

        let lastIndex = 0;
        const newlineIndexes = [];
        indexes.forEach((blockIndex, index) => {
            const text = htmlString.substring(lastIndex, blockIndex - 7);
            lastIndex = blockIndex;
            if (text.trim().length === 0 && !text.match(/\n/)) {
                return;
            }

            // if we have newline of quote or we are in last element not map the sting to newline
            if (text.match(/[\n|>][>]?[\s]?$/) || index === indexes.length - 1) {
                joinedText += text;
            } else {
                newlineIndexes.push(blockIndex - 7);
                joinedText += `${text}\n`;
            }
        });
        if (lastIndex < htmlString.length) {
            const text = htmlString.substring(lastIndex, htmlString.length).replace(/\[block\]/g, '');
            joinedText += (text.trim().length === 0 && !text.match(/\n/)) || lastIndex === 0 ? text : `\n${text}`;
        }

        return joinedText;
    }
```

Reply:
addressed below.